README.md
@tru_id/cli-plugin-dev-server
=============================

Runs a development server from the tru.ID CLI

[![oclif](https://img.shields.io/badge/cli-oclif-brightgreen.svg)](https://oclif.io)
[![Version](https://img.shields.io/npm/v/@tru_id/cli-plugin-dev-server.svg)](https://npmjs.org/package/@tru_id/cli-plugin-dev-server)
[![Downloads/week](https://img.shields.io/npm/dw/@tru_id/cli-plugin-dev-server.svg)](https://npmjs.org/package/@tru_id/cli-plugin-dev-server)
[![License](https://img.shields.io/npm/l/@tru_id/cli-plugin-dev-server.svg)](https://github.com/tru-ID/cli-plugin-dev-server/blob/master/package.json)

<!-- toc -->
* [Installation](#installation)
* [Commands](#commands)
<!-- tocstop -->

# Installation

To install plugins you'll need the [tru.ID CLI](https://github.com/tru-ID/cli) `v0.8.0` or above.

```
npm install -g @tru_id/cli@canary
```

You can then install the plugin using:

```sh-session
$ tru plugins:install @tru_id/cli-plugin-dev-server
```

# Commands
<!-- commands -->
* [`tru server`](#tru-server)

## `tru server`

Runs a development server from the CLI

```
USAGE
  $ tru server

OPTIONS
  -h, --help                                         show CLI help
  -p, --port=port                                    [default: 8080] The port the server should listen on
  -s, --localtunnel-subdomain=localtunnel-subdomain  The subdomain for the local tunnel
  -t, --localtunnel                                  Run a localtunnel to expose the server to the Internet

  --project-dir=project-dir                          [default: .] The directory that contains the tru.json Project
                                                     configuration file

EXAMPLES
  $ tru server

  $ tru server -t -p 4000
```

_See code: [src/commands/server.ts](https://github.com/tru-ID/cli-plugin-dev-server/blob/v0.0.7/src/commands/server.ts)_
<!-- commandsstop -->

## Releasing

### CHANGELOG

Ensure the `CHANGELOG.md` is updated:

```
yarn run changelog
```

Finesse the contents of `CHANGELOG.md` and commit.

### Tag

Tag the version that is being released with the version of the package and, optionally, the tag to be used in NPM:

```
git tag v{version}@{tag}
```

Push the update to the remote.

```
git push v{version}@{tag}
```

### Publish to NPM

To release version run:

```
yarn run publish:latest
```

To release a canary version only run:

```
yarn run publish:canary
```

## License

[MIT](LICENSE)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.8](https://github.com/tru-ID/cli-plugin-dev-server/compare/v0.0.7...v0.0.8) (2021-06-21)

- bump dependency on [@tru_id/dev-server](https://github.com/tru-ID/server-example-node) to v0.2.3
- Added POST /traces debug route

### [0.0.6](https://github.com/tru-ID/cli-plugin-dev-server/compare/v0.0.4...v0.0.6) (2021-06-01)

- bump dependency on [@tru_id/dev-server](https://github.com/tru-ID/server-example-node) to v0.2.2

### [0.0.4](https://github.com/tru-ID/cli-plugin-dev-server/compare/v0.0.3...v0.0.4) (2021-03-11)

- bump dependency on [@tru_id/dev-server](https://github.com/tru-ID/server-example-node) to v0.1.6 for `/my-ip` endpoint support

### [0.0.3](https://github.com/tru-ID/cli-plugin-dev-server/compare/v0.0.2...v0.0.3) (2021-03-09)

- Update the `server` command description

### 0.0.2 (2021-02-05)

- bump dependency on [@tru_id/dev-server](https://github.com/tru-ID/server-example-node) to v0.1.4

### Bug Fixes

- remove augmenting process.env for config ([b8480ed](https://github.com/tru-ID/cli-plugin-dev-server/commit/b8480ed4f37d3e5ecba142b2da18a69db9dc6e89))

### 0.0.1 (2021-02-04)

### Features

- `server` command ([32a5480](https://github.com/tru-ID/cli-plugin-dev-server/commit/32a5480f603dee42d7229142e298f18498382577))


test/commands/server.test.ts
import {expect, test} from '@oclif/test'

describe('server', () => {
  // test
  // .stdout()
  // .command(['server', '--help'])
  // .it('runs server -h', ctx => {
  //   expect(ctx.stdout).to.contain('tru server')
  // })
})


src/commands/server.ts
import * as path from 'path'
import * as fs from 'fs-extra'

import { Command, flags } from '@oclif/command'
const server = require('@tru_id/dev-server')

export default class Serve extends Command {
  static description = 'Runs a development server from the CLI'

  static examples = [
    `$ tru server
`,
    `$ tru server -t -p 4000
`,
  ]

  static flags = {
    help: flags.help({ char: 'h' }),
    localtunnel: flags.boolean({
      description: 'Run a localtunnel to expose the server to the Internet',
      char: 't',
      default: false,
    }),
    'localtunnel-subdomain': flags.string({
      description: 'The subdomain for the local tunnel',
      char: 's',
    }),
    // TODO share with @tru_id/cli
    'project-dir': flags.string({
      description: 'The directory that contains the tru.json Project configuration file',
      default: '.',
    }),
    port: flags.integer({
      description: 'The port the server should listen on',
      char: 'p',
      default: 8080,
    }),
  }

  async run() {
    const { flags } = this.parse(Serve)
    const projectConfig = await this.loadProjectConfig(flags)

    server.serve({
      port: flags.port,
      localtunnel: {
        enabled: flags.localtunnel,
        subdomain: flags['localtunnel-subdomain'],
      },
      project: {
        client_id: projectConfig.credentials[0].client_id,
        client_secret: projectConfig.credentials[0].client_secret,
      },
    })
  }

  // TODO share with @tru_id/cli
  async loadProjectConfig(flags: any) {
    const projectDirectory = flags['project-dir'] ?? process.cwd()
    const projectConfigFullPath = path.join(projectDirectory, 'tru.json')
    const projectConfigExists = fs.existsSync(projectConfigFullPath)
    if (projectConfigExists === false) {
      this.log(`A project configuration file does not exist at "${projectConfigFullPath}".\n` +
        'Please provide a valid directory path using the `--project-dir` flag or run `tru projects:create` to create a project and associated configuration file.')
      this.exit(1)
    }
    try {
      return await fs.readJson(projectConfigFullPath)
    }
    catch (error) {
      this.log('There was a problem loading the tru.json configuration file',
        `${error.toString()} ${JSON.stringify(error)}`)
      this.exit(1)
    }
  }
}
